Gambot.Modules.Roll/DiceEvaluator.cs

```diff
@@ -0,0 +1,271 @@
+
+        private string _Symbols = "khld*/+-()";
+        private IEnumerable<DiceToken> Tokenize(string expression)
+        {
+            List<DiceToken> tokens = new List<DiceToken>();
+            var expr = expression.ToLower();
+
+            var i = 0;
+            while (i < expr.Length)
+            {
+                if (_Symbols.Contains(expr[i]))
+                {
+                    var token = new DiceToken() { IsNumber = false, StringValue = expr[i].ToString() };
+                    if (tokens.Count > 0)
+                    {
+                        token.PrecedingToken = tokens[tokens.Count - 1];
+                        tokens[tokens.Count - 1].ProceedingToken = token;
+                    }
+                    tokens.Add(token);
+                }
+                else if (Char.IsDigit(expr[i]))
+                {
+                    var start = i;
+                    do { ++i; } while (i < expr.Length && Char.IsDigit(expr[i]));
+
+                    var token = new DiceToken() { IsNumber = true, IntValue = Convert.ToInt32(expr.Substring(start, i - start)) };
+                    if (tokens.Count > 0)
+                    {
+                        token.PrecedingToken = tokens[tokens.Count - 1];
+                        tokens[tokens.Count - 1].ProceedingToken = token;
+                    }
+                    tokens.Add(token);
+                    continue;
+                }
+                ++i;
+            }
+
+            var tmp = new List<DiceToken>();
+            foreach (var token in tokens)
+            {
+                if (token.StringValue == "d" && !(token.PrecedingToken != null && token.PrecedingToken.IsNumber))
+                {
+                    tmp.Add(new DiceToken() { IsNumber = true, IntValue = 1 });
+                }
+
+                tmp.Add(token);
+
+                if ((token.StringValue == "h" || token.StringValue == "k" || token.StringValue == "l") && !(token.ProceedingToken != null && token.ProceedingToken.IsNumber))
+                {
+                    tmp.Add(new DiceToken() { IsNumber = true, IntValue = 1 });
+                }
+            }
+            tokens.Clear();
+
+            for (var tmpIdx = 0; tmpIdx < tmp.Count(); ++tmpIdx)
+            {
+                tmp[tmpIdx].ProceedingToken = null;
+                tmp[tmpIdx].PrecedingToken = (tmpIdx == 0) ? null : tmp[tmpIdx - 1];
+
+                if (tmpIdx > 0)
+                {
+                    tmp[tmpIdx - 1].ProceedingToken = tmp[tmpIdx];
+                }
+            }
+
+            return tmp;
+        }
+
+        private bool IsHigherPrecedence(string first, string second)
+        {
+            var precString =_Symbols.Substring(0, _Symbols.Length - 2);
+            var firstPrec = precString.IndexOf(first);
+            var secondPrec = precString.IndexOf(second);
+            return firstPrec <= secondPrec;
+        }
+
+        private int NumberOfArgumentsForSymbol(string symbol)
```

Comment:
Piggybacking off of two of my previous comments, it may make sense to refactor this to be a bit more OOP. Instead of having `IsHigherPrecedence(symbol)` and `NumberOfArgumentsForSymbol(symbol)`, it would be more idiomatic to have some sort of class for the symbols, so you could then do `symbol1.Precedence > symbol2.Precedence` and `symbol.NumberOfArguments`, respectively